stock_picking_group_by_partner_by_carrier/models/stock_picking.py

```diff
@@ -171,3 +171,38 @@ def get_customer_refs(self):
             move_lines = self.move_lines
         references = move_lines.mapped("sale_line_id.order_id.client_order_ref")
         return set(filter(None, references))
+
+    def demand_is_completed(self):
+        """Return whether the picking completes the demand
+
+        If the outgoing picking comes from a sale order, it completes the
+        demand if the order have been completed. If doesn't come from a sale
+        order, a picking completes the order if it is the last one from those
+        needed to fulfill an order.
+        """
+        self.ensure_one()
+
+        sales = self.move_line_ids.mapped("move_id.sale_line_id.order_id")
```

Comment:
can't we get to orders via procurement group?

Reply:
Yes, thanks.